plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSExpressionValidator.xtend

```diff
@@ -696,6 +697,24 @@ class N4JSExpressionValidator extends AbstractN4JSDeclarativeValidator {
 					}
 				}
 			}
+			
+			val rhs = relationalExpression.rhs;
+			if (rhs instanceof UnaryExpression) {
+				if ((rhs as UnaryExpression).getOp().equals(UnaryOperator.INV)) {
+					val innerExpression = (rhs as UnaryExpression).expression;
+					var rhsTypeRef = ts.tau(innerExpression);
+					
+					if (innerExpression instanceof UnaryExpression) {
+						rhsTypeRef = ts.tau((innerExpression as UnaryExpression).expression);
+					}
```

Comment:
First check the expression, only then compute tau. At the moment the computation in line 705 is not needed if the expression is an instance of UnaryExpression. In xtend, you can even write

```
val rhsTypeRef = tau (
        if (innerExpression instanceof UnaryExpression) {
           (innerExpression as UnaryExpression).expression;
        } else {
            innerExpression
        });
```

Reply:
Another remark: You check for unary expression in general. I'm not quite sure if this is not too general since we only want to handle the INV case here. I wasn't able to create a problem in N4JS, probably because the ExpectedTypeJudgment is handling this correct. I would suggest to only really limit the extra handling to the combination of INSTANCEOF and INV though.